### PR TITLE
async destination -- lock drain to

### PR DESCRIPTION
**Pull request by cgardens**
## What
* allow draining from the queue to happen in a lock to prevent interleaved reads from different workers.
* took the implementation approach from how `LinkedBlockingQueue` does it by using a `ReentrantLock`.

### CODE REVIEW

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/BufferManager.java [hunk 1]
```diff
@@ -154,34 +157,47 @@ public Optional<Instant> getTimeOfLastRecord(final StreamDescriptor streamDescri
...
+          final long newSize = memoryItem.size() + bytesRead.get();
+          if (newSize <= bytesToRead) {
+            bytesRead.addAndGet(memoryItem.size());
+            return true;
+          } else {
+            return false;
+          }
+        }).map(MemoryItem::item)
+            .toList()
+            .stream();
```
Comment (cgardens) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/BufferManager.java:191]: hack for now. we can switch this to a for loop, but didn't want to slow us down on this right now.

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/BufferManager.java [hunk 2]
```diff
@@ -120,11 +121,13 @@ static class BufferManagerDequeue {
 
     private final GlobalMemoryManager memoryManager;
     private final Map<StreamDescriptor, MemoryBoundedLinkedBlockingQueue<AirbyteMessage>> buffers;
+    private final Map<StreamDescriptor, ReentrantLock> bufferLocks;
 
     public BufferManagerDequeue(final GlobalMemoryManager memoryManager,
                                 final Map<StreamDescriptor, MemoryBoundedLinkedBlockingQueue<AirbyteMessage>> buffers) {
       this.memoryManager = memoryManager;
       this.buffers = buffers;
+      bufferLocks = new HashMap<>();
```
Comment (ryankfu) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/BufferManager.java:130]: Does this need to be a `ConcurrentHashMap`?

Comment (cgardens) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/BufferManager.java:130]: i think actually it probably should. 2 threads could try to put the lock at once.

Comment (cgardens) [reply airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/BufferManager.java:130]: buffers should also be a concurrenthasmap

**Review by ryankfu — APPROVED** [review]
Makes sense pending updates to `ConcurrentHashMap`

**Pull request merged**